yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala

```diff
@@ -616,17 +620,29 @@ private[spark] class Client(
    * throw an appropriate SparkException.
    */
   def run(): Unit = {
-    val (yarnApplicationState, finalApplicationStatus) = monitorApplication(submitApplication())
-    if (yarnApplicationState == YarnApplicationState.FAILED ||
-      finalApplicationStatus == FinalApplicationStatus.FAILED) {
-      throw new SparkException("Application finished with failed status")
-    }
-    if (yarnApplicationState == YarnApplicationState.KILLED ||
-      finalApplicationStatus == FinalApplicationStatus.KILLED) {
-      throw new SparkException("Application is killed")
-    }
-    if (finalApplicationStatus == FinalApplicationStatus.UNDEFINED) {
-      throw new SparkException("The final status of application is undefined")
+    val appId = submitApplication()
+    if (fireAndForget) {
+      logInfo("... polling ResourceManager for application state")
```

Comment:
If take a look back at @tgravescs 's comments, you will find out that he make same propose before. The problem is using `monitorApplication(appId, returnOnRunning = true)` doesn't totally equal to what `fire-and-forget` means.